src/components/pages/dashboard/panels/map/mapPanel.js

```diff
@@ -14,73 +15,133 @@ import {
+    // });
   }
 
   componentWillReceiveProps(nextProps) {
+    if (!this.map && nextProps.azureMapsKey) {
+      this.initializeMap(nextProps.azureMapsKey);
+    }
     this.calculatePins(nextProps);
   }
 
-  calculatePins(props, initialCall = false) {
+  initializeMap(azureMapsKey) {
+    const seattlePosition = new AzureMaps.data.Position(-122.3320708, 47.6062);
+    this.map = new AzureMaps.Map('map', {
+      'subscription-key': azureMapsKey,
+      center: seattlePosition,
+      zoom: 11
+    });
+
+    this.map.addPins([], {
+      name: nominalDeviceLayer,
+      cluster: true,
+      icon: 'pin-blue'
+    });
+
+    this.map.addPins([], {
+      name: warningDevicesLayer,
+      cluster: true,
+      icon: 'pin-darkblue'
+    });
+
+    this.map.addPins([], {
+      name: criticalDevicesLayer,
+      cluster: true,
+      icon: 'pin-red'
+    });
+  }
+
+  calculatePins(props, mounting = false) {
     const deviceIds = Object.keys(props.devices);
-    if (!initialCall && deviceIds.join() === Object.keys(this.props.devices).join()) return;
-
-    const devicePins = deviceIds
-      .map(key => props.devices[key])
-      .filter(({ properties }) => properties.latitude && properties.longitude)
-      .map(({ id, properties }) =>
-        new atlas.data.Feature(new atlas.data.Point([properties.longitude, properties.latitude]), {
-          name: id,
-          address: properties.location || ''
-        })
-      );
-
-    if (this.map && devicePins.length) {
-      this.map.addPins(devicePins, {
-        name: deviceLayer
-      });
-
-      const lons = [];
-      const lats = [];
-      devicePins.forEach(({ geometry: { coordinates: [ longitude, latitude ] } }) => {
-        lons.push(longitude);
-        lats.push(latitude);
-      });
-
-      const swLon = Math.min.apply(null, lons);
-      const swLat = Math.min.apply(null, lats);
-      const neLon = Math.max.apply(null, lons);
-      const neLat = Math.max.apply(null, lats);
-
-      // Zoom the map to the bounding box of the devices
-      this.map.setCameraBounds({
-        bounds: [swLon, swLat, neLon, neLat],
-        padding: 50
-      });
+    const prevDeviceIds = Object.keys(this.props.devices);
+    const boundZoomToDevices =
+      (deviceIds.length > 0 && prevDeviceIds.length === 0)
+      || (props.azureMapsKey && !this.props.azureMapsKey)
+      || (mounting && props.azureMapsKey && deviceIds.length > 0);
+
+    if (this.map && deviceIds.length > 0) {
+      const geoLocatedDevices = this.extractGeoLocatedDevices(props.devices);
+
+      const { normal, warning, critical }  = this.devicesToPins(geoLocatedDevices, props.devicesInAlarm)
+
+      if (this.map && (normal.length || warning.length || critical.length)) {
+        this.map.addPins(normal, { name: nominalDeviceLayer, overwrite: true });
+        this.map.addPins(warning, { name: warningDevicesLayer, overwrite: true });
+        this.map.addPins(critical, { name: criticalDevicesLayer, overwrite: true });
```

Comment:
The || condition will execute even when only one variable has some value. Do you want to add pins if lets say `warning.length == 0` and `normal.length != 0`

Reply:
Actually, we don't need this check at all. I removed it.